### PR TITLE
Support vectorized gradients in Nx.Defn.grad

### LINKED ISSUE #1533: Nx.Defn.Grad returns sum of gradient over vectorized axis?
Thanks for making Nx!

I tried to use `value_and_grad` on a function that takes two inputs: a vectorized tensor and a non-vectorized tensor.

``` elixir
defmodule Foo do
  import Nx.Defn
  defn f(x, y) do
    x + y
  end

  defn f_and_grad(x, y) do
    value_and_grad(y, fn y -> Foo.f(x, y) end)
  end
end

x = ~VEC[0 1] |> vectorize(:bar)
Foo.f_and_grad(x, 1)
```

This evaluates to:

```
{#Nx.Tensor<
   vectorized[bar: 2]
   s64
   EXLA.Backend<host:0, 0.731981912.321781778.128426>
   [1, 2]
 >,
 #Nx.Tensor<
   f32
   EXLA.Backend<host:0, 0.731981912.321781778.128427>
   2.0
 >}
```

The value is correct and maintains the vectorized axis of the vectorized input to `x`, but the gradient surprises me. I would have expected a vectorized tensor rank-1 dimension-2 vector with the same `:foo` axis and which is everywhere `1`; it looks like instead Nx is summing up the two gradients.

Is this behavior expected? If so, is there any way to make Nx return a vectorized gradient?

Thanks!

### PR DESCRIPTION
**Pull request by blasphemetheus**
fixes #1533

## Summary

Makes `Nx.Defn.grad` work correctly with vectorized tensors. The grad system now re-vectorizes expression tree nodes during the backward pass and adjusts per-op options (axes, padding, strides, etc.) to account for vectorized dimensions.

Key changes:

  - **Re-vectorization in backward pass**: `recur_to_grad` tracks vectorized axis names through `parents_tree` and re-vectorizes args/ans during gradient computation. `adjust_vectorized_args` handles ~20 ops.

  - **apply_vectorized boundaries**: Operations like `cumulative_sum`, `triangular_solve`, etc. use `apply_vectorized` internally, creating a devectorized region in the expression tree. The grad system detects these boundaries (in `parents_args(:optional)` and `update_grads(:optional)`) and stops propagating vectorized names / devectorizes gradients at the boundary.

  - **Cond fix**: Vectorized `cond` creates separate cond_then/cond_else nodes sharing the same to_grad variables. Prevents cross-contamination by cleaning to_grad entries before each clause's inner computation.

  - **Conv fix**: When conv collapses the vectorized dim into `batch_group_size`, devectorizes args/ans in `grad_conv` since the vectorized dim is already in the batch.

  - **Batched linalg helpers**: QR, Cholesky, and triangular_solve grads now use batch-aware matmul helpers for 3D+ tensors.

  - **Concrete tensor handling**: `parents_args` and `Expr.parameter` handle non-Expr tensors captured in grad closures.

  - **`assert_all_close` fix**: Test helper now devectorizes before comparing.

  ## Test plan

  - 346 tests pass (118 new), 0 failures
  - `vectorized_grad_test.exs`: 91 correctness tests comparing vectorized grad vs per-element grad across all supported ops
  - `grad_test.exs`: 27 edge case tests (vectorize/devectorize inside grad body, non-vectorized target, error messages, window_scatter, second-order grad, tuple grad)
  - Full Nx suite: 2,692 tests pass

### CODE REVIEW

**Comment by blasphemetheus** [comment]
gather issue documented at #1698

**Comment by blasphemetheus** [comment]
failure is a doctest with 1 different difference on a tensor, so should be clean. Any other tests that would be good to have?

**Review by josevalim — COMMENTED** [review]
I wonder if there is a way to do holistically? Like we remove vectorization before grading and then reapply it afterwards, or similar. Or the opposite, we make sure we vectorize upfront so all operations inside a properly vectorized... unsure.

File: nx/lib/nx/defn/grad.ex [hunk 1]
```diff
@@ -214,7 +214,22 @@ defmodule Nx.Defn.Grad do
...
+          Nx.broadcast(res, arg)
+
+        {vectorized_axes, []} ->
+          devec_arg = Nx.devectorize(arg, keep_names: true)
+          res = Nx.broadcast(res, devec_arg)
+          Nx.vectorize(res, vectorized_axes)
+
+        {_, _} ->
+          Nx.broadcast(res, arg)
+      end
```
Comment (polvalente) [inline nx/lib/nx/defn/grad.ex]: Is this really needed?

File: nx/lib/nx/defn/grad.ex [hunk 2]
```diff
@@ -285,6 +311,170 @@ defmodule Nx.Defn.Grad do
...
+    [x, adjusted_dims, adjusted_opts]
+  end
+
+  # fft/ifft: [t, opts] — length refers to last axis, needs no axis adjustment
+  # but Nx.rank(t) is used in the grad clause (returns inner rank, which is correct)
+  defp adjust_vectorized_args(:fft, args, _offset), do: args
+  defp adjust_vectorized_args(:ifft, args, _offset), do: args
+
+  # conv: complex, skip for now
+  defp adjust_vectorized_args(:conv, args, _offset), do: args
```
Comment (polvalente) [inline nx/lib/nx/defn/grad.ex]: This should probably raise with unsupported

File: nx/lib/nx/defn/grad.ex [hunk 3]
```diff
@@ -285,6 +311,170 @@ defmodule Nx.Defn.Grad do
...
+  # Default: no adjustment needed (elementwise ops, etc.)
+  defp adjust_vectorized_args(_op, args, _offset), do: args
+
+  # Helpers for adjust_vectorized_args
+
+  defp offset_axes(axes, offset) when is_list(axes) do
+    axes
+    |> Enum.map(&(&1 - offset))
+    |> Enum.filter(&(&1 >= 0))
+  end
```
[the remark this answers is not in the document]
Comment (polvalente) [reply nx/lib/nx/defn/grad.ex]: Actually, if we reach negative values, something is very wrong, I think

**Comment by polvalente** [comment]
Mixed vectorization should work because computations work on that scenario. We even have things such as reshape_vectors and broacast_vectors for dealing with mixed vectors scenarios.

Also, not all vectorized tensors have the same vectorized axes

File: nx/lib/nx/defn/grad.ex (hunk 1 above)
Comment (blasphemetheus) [reply nx/lib/nx/defn/grad.ex]: replaced with Nx.broadcast_vectors

File: nx/lib/nx/defn/grad.ex (hunk 2 above)
Comment (blasphemetheus) [reply nx/lib/nx/defn/grad.ex]: oh absolutely, ArgumentError now

File: nx/lib/nx/defn/grad.ex [hunk 4]
```diff
@@ -285,6 +311,170 @@ defmodule Nx.Defn.Grad do
...
+      |> adjust_keyword_drop(:window_dilations, offset)
+      |> adjust_keyword_padding(:padding, offset)
+
+    [x, adjusted_dims, adjusted_opts]
+  end
+
+  # fft/ifft: [t, opts] — length refers to last axis, needs no axis adjustment
+  # but Nx.rank(t) is used in the grad clause (returns inner rank, which is correct)
+  defp adjust_vectorized_args(:fft, args, _offset), do: args
+  defp adjust_vectorized_args(:ifft, args, _offset), do: args
```
[the remark this answers is not in the document]
Comment (blasphemetheus) [reply nx/lib/nx/defn/grad.ex]: aaand removed em :) following feedback on another PR (never gonna be used basically)

**Review by polvalente — COMMENTED** [review]
Do we have a test where vectorize/devectorize/revectorize are used inside the grad function itself?

We're improving support for inputs with gradients, but what happens when the vectors are not observable from the output?

**Comment by josevalim** [comment]
I don’t think we should worry about using vectorize inside grad. We can place it as a restriction for now.

File: nx/test/support/helpers.ex [hunk 5]
```diff
@@ -23,8 +23,15 @@ defmodule Nx.Helpers do
     atol = opts[:atol] || 1.0e-4
     rtol = opts[:rtol] || 1.0e-4
 
-    unless Nx.all_close(lhs, rhs, atol: atol, rtol: rtol, equal_nan: opts[:equal_nan]) ==
-             Nx.tensor(1, type: {:u, 8}) do
+    close =
```
[the remark this answers is not in the document]
Comment (blasphemetheus) [reply nx/test/support/helpers.ex]: these changes are intended to match the approach used in `Nx.Testing.assert_all_close` (public API version)

**Comment by blasphemetheus** [comment]
Reduced tests a bit, squashed into 3 commits. Passing on all tests, no tests left skipped for vectorized grad

File: nx/lib/nx/defn/expr.ex [hunk 6]
```diff
@@ -68,6 +68,13 @@ defmodule Nx.Defn.Expr do
     parameter(tensor, context, pos)
   end
 
+  def parameter(%T{} = tensor, pos) do
+    # Concrete (non-Expr) tensor — convert to constant expression.
+    # This happens when captured tensors end up in optional args
+    # during grad callback re-evaluation.
+    to_expr(tensor) |> then(&parameter(&1, pos))
```
Comment (polvalente) [inline nx/lib/nx/defn/expr.ex:75]:
```suggestion
    tensor |> to_expr(tensor) |> parameter(pos)
```

Although @josevalim will probably be able to say if this is indeed a change we should be making

File: nx/lib/nx/lin_alg/cholesky.ex [hunk 7]
```diff
@@ -76,8 +76,8 @@ defmodule Nx.LinAlg.Cholesky do
   end
 
   defn cholesky_grad(l, _input, g) do
-    num = g |> Nx.tril() |> Nx.dot([0], l, [0]) |> Nx.transpose()
-    den = l |> Nx.shape() |> Nx.eye() |> Nx.add(1)
+    num = g |> Nx.tril() |> matmul_contract_rows(l) |> batch_transpose()
```
Comment (polvalente) [inline nx/lib/nx/lin_alg/cholesky.ex:79]:
```suggestion
    num = g |> Nx.tril() |> dot_contract_rows(l) |> batch_transpose()
```

we don't use matmul in the Nx codebase

Comment (polvalente) [reply nx/lib/nx/lin_alg/cholesky.ex:79]: Also, if we're transposing the result:

what we have is `transpose(num) = transpose(tril(g)) dot l` which is equal to `num = l dot tril(g)`, so I think we can just do `num = Nx.dot(l, Nx.tril(g))`

File: nx/lib/nx/lin_alg/cholesky.ex [hunk 8]
```diff
@@ -87,16 +87,43 @@ defmodule Nx.LinAlg.Cholesky do
...
   end
 
+  deftransformp matmul_contract_rows(a, b) do
+    rank = tuple_size(a.shape)
+
+    if rank <= 2 do
+      Nx.dot(a, [0], b, [0])
+    else
+      batch_axes = Enum.to_list(0..(rank - 3))
+      Nx.dot(a, [-2], batch_axes, b, [-2], batch_axes)
```
Comment (polvalente) [inline nx/lib/nx/lin_alg/cholesky.ex:100]: using `Nx.dot(a, [-2], b, [-2])` should be equivalent in all cases, I think.

Comment (polvalente) [reply nx/lib/nx/lin_alg/cholesky.ex:100]: But, see my comment about rewriting it altogether

File: nx/lib/nx/lin_alg/cholesky.ex [hunk 9]
```diff
@@ -87,16 +87,43 @@ defmodule Nx.LinAlg.Cholesky do
...
+  deftransformp eye_like(l) do
+    rank = tuple_size(l.shape)
+    n = elem(l.shape, rank - 1)
+
+    if rank <= 2 do
+      Nx.eye(n)
+    else
+      batch_dims = l.shape |> Tuple.to_list() |> Enum.take(rank - 2)
+      Nx.eye(n) |> Nx.broadcast(List.to_tuple(batch_dims ++ [n, n]))
+    end
```
Comment (polvalente) [inline nx/lib/nx/lin_alg/cholesky.ex:124]:
```suggestion
    shape = put_elem(l.shape, rank - 2, n)
    Nx.eye(shape)
```

Comment (polvalente) [reply nx/lib/nx/lin_alg/cholesky.ex:124]: and since this code is small we can put it back instead of having a defp

File: nx/lib/nx/lin_alg/qr.ex [hunk 10]
```diff
@@ -151,13 +151,26 @@ defmodule Nx.LinAlg.QR do
     # Equation (3)
     r_inv = Nx.LinAlg.invert(r)
 
-    m = Nx.dot(r, Nx.LinAlg.adjoint(dr)) |> Nx.subtract(Nx.dot(Nx.LinAlg.adjoint(dq), q))
+    m =
+      matmul(r, Nx.LinAlg.adjoint(dr))
```
Comment (polvalente) [inline nx/lib/nx/lin_alg/qr.ex:155]: same consideration about matmul

File: nx/lib/nx/lin_alg/qr.ex [hunk 11]
```diff
@@ -151,13 +151,26 @@ defmodule Nx.LinAlg.QR do
...
   end
+
+  deftransformp matmul(a, b) do
+    rank = tuple_size(a.shape)
+
+    if rank <= 2 do
+      Nx.dot(a, b)
+    else
+      batch_axes = Enum.to_list(0..(rank - 3))
+      Nx.dot(a, [-1], batch_axes, b, [-2], batch_axes)
```
Comment (polvalente) [inline nx/lib/nx/lin_alg/qr.ex:173]: This code assumes both sides have the same rank

Comment (polvalente) [reply nx/lib/nx/lin_alg/qr.ex:173]: Let's just use it in-place because it relies on this implied condition

File: nx/test/support/helpers.ex [hunk 12]
```diff
@@ -23,8 +23,15 @@ defmodule Nx.Helpers do
...
+      lhs
+      |> Nx.all_close(rhs, atol: atol, rtol: rtol, equal_nan: opts[:equal_nan])
+      |> Nx.devectorize(keep_names: false)
+      |> Nx.backend_transfer(Nx.BinaryBackend)
+      |> Nx.to_flat_list()
+      |> Enum.all?(&(&1 == 1))
+
+    unless close do
```
Comment (polvalente) [inline nx/test/support/helpers.ex:34]: We should probably be fixing Nx.Testing and using it here instead

File: nx/test/nx/defn/vectorized_grad_test.exs [hunk 13]
```diff
@@ -0,0 +1,808 @@
+defmodule Nx.Defn.VectorizedGradTest do
```
Comment (polvalente) [inline nx/test/nx/defn/vectorized_grad_test.exs:1]: Not sure if this should be a separate file

File: nx/lib/nx/defn/grad.ex [hunk 14]
```diff
@@ -167,13 +177,19 @@ defmodule Nx.Defn.Grad do
...
 
-        recur_parents_tree(arg, {parents, nodes}, arg_vectorized_names)
+        %Nx.Defn.Expr{id: arg_id} ->
+          arg_vectorized_names = compute_arg_vectorized_names(t, parent_vectorized_names)
+          parents = Map.update(parents, arg_id, [parent_id], &[parent_id | &1])
+
+          recur_parents_tree(arg, {parents, nodes}, arg_vectorized_names)
+
+        _concrete ->
+          # Concrete (non-Expr) tensor captured in closure — skip tree traversal
```
Comment (polvalente) [inline nx/lib/nx/defn/grad.ex:191]: @josevalim should this be happening here?

File: nx/lib/nx/defn/grad.ex [hunk 15]
```diff
@@ -209,16 +225,85 @@ defmodule Nx.Defn.Grad do
...
+            reraise ArgumentError,
+                    "cannot compute gradient: a vectorized axis name was reused with a " <>
+                      "different size inside the grad function. Redefining the meaning of " <>
+                      "a vectorized axis (e.g., devectorize then reshape then re-vectorize " <>
+                      "with the same name at a different size) is not supported. " <>
+                      "Original error: #{Exception.message(e)}",
+                    __STACKTRACE__
+          else
+            reraise e, __STACKTRACE__
+          end
```
Comment (polvalente) [inline nx/lib/nx/defn/grad.ex:246]: This shouldn't be captured here. At best, we should capture closer to where the error happens.

File: nx/lib/nx/defn/grad.ex [hunk 16]
```diff
@@ -209,16 +225,85 @@ defmodule Nx.Defn.Grad do
...
+      cond do
+        # Extra inner dims from devectorize baking batch into shape —
+        # extract each batch element's own gradient (the diagonal).
+        res.vectorized_axes == arg.vectorized_axes and
+            tuple_size(res.shape) > tuple_size(arg.shape) ->
+          batch_size = elem(res.shape, 0)
+
+          idx =
+            Nx.iota({batch_size}, type: :s32)
+            |> Nx.vectorize(arg.vectorized_axes)
```
Comment (polvalente) [inline nx/lib/nx/defn/grad.ex:263]: I think this is overfitting a very specific case. Can you walk me through the logic?
This code only works if length(arg.vectorized_axes) is 0 or 1

File: nx/lib/nx/defn/grad.ex [hunk 17]
```diff
@@ -209,16 +225,85 @@ defmodule Nx.Defn.Grad do
...
+    # Mixed vectorized/non-vectorized terms from apply_vectorized —
+    # devectorize all before summing to avoid broadcast_vectors collision
+    has_vec = Enum.any?(gs, &(&1.vectorized_axes != []))
+    has_no_vec = Enum.any?(gs, &(&1.vectorized_axes == []))
+
+    gs =
+      if has_vec and has_no_vec do
+        Enum.map(gs, &Nx.devectorize(&1, keep_names: false))
+      else
+        gs
```
Comment (polvalente) [inline nx/lib/nx/defn/grad.ex:302]: what's the case where we don't want devectorize to happen?

**Comment by polvalente** [comment]
@blasphemetheus after some discussion with @josevalim, we think this is the wrong approach overall.

We think that the solution should be akin to a thin wrapper on `Nx.Defn.Grad.transform` instead, where we deal with the vectors in the input and output steps, without leaking them downwards, kinda like what Nx does with passing tensors to backends.

This approach can include a requirement for functions not revectorizing stuff internally, I think.

**Comment by blasphemetheus** [comment]
thanks for the review! Hopefully the tests are useful here.
